source/common/http/conn_manager_impl.cc

```diff
@@ -1553,6 +1554,10 @@ bool ConnectionManagerImpl::ActiveStreamFilterBase::commonHandleAfterHeadersCall
   if (status == FilterHeadersStatus::StopIteration) {
     stopped_ = true;
     return false;
+  } else if (status == FilterHeadersStatus::StopAllTypesIteration) {
+    stopped_ = true;
+    stopped_all_ = true;
```

Comment:
Where do we look at `stopped_all_` to determine to not call any filter callbacks? I don't see that? This is the place also that for data we will have to decide on buffer vs watermark processing?

Reply:
I checked stopped_all_ value in line 1597 and line 1623 in conn_manager_impl.cc. Let's say there are 3 filters:

filter1  ----- filter2 ----- filter3

where filter2::decodeHeaders() returns StopAllIteration. We still allow data to call filter1::decodeData(), and filter2::decodeData(), but will not call filter3::decodeData(). This is achieved by the check filter2::stop_all_ value in line 1597. If filter2::stop_all_ is true, we will buffer data but not call commonContinue(), and because commonHandleAfterDataCallback returns false, filter3::decodeData() will not be called. 

The same logic applies to trailers, where stop_all_ value is checked in line 1623. But I can be completely wrong. Please let me know the feedback. Thanks!

Reply:
OK I see. I think though we don't want to call any other filter2 methods, including data and trailers. This is how most filters would use this I think, including auth, rate limit, etc. where they make an outbound call as part of the decodeHeaders() path. Right?

Assuming ^ is what we want, I think this means that you need to check stop all status before calling filter functions, and then also handle buffer/watermark for data?

Reply:
Ah, I see! I mistakenly thought we didn't allow callbacks for filters after filter2, excluding filter2. Will make the change. Thanks!

About stop-all options for data. I thought we want stop-all for headers only at least for now? (https://github.com/envoyproxy/envoy/issues/5842#issuecomment-461550426).  But it makes sense to be consistent with data as well. Will make the change and address other comments as well. Thanks for the feedback!

Reply:
> About stop-all options for data. I thought we want stop-all for headers only at least for now?

We do want it only for headers, but what I'm saying is that we need to tell the HCM how to handle subsequent data. Do we buffer up to the limit and then error? Or do we buffer and try to apply flow control? Those are the two options.

Reply:
Sorry I misread the comment!

My original thoughts was if users don't want to buffer large data or want the flow control feature, they may add a rate-limiting filter before the filter that can return StopAllIteration for headers: 

"// To avoid buffering large amount of data, a rate-limiting filter can be added before the filter that 
// can return StopAllTypesIteration."

If users don't care, we can buffer up to the limit and then error. But I think I missed to set decoder_filters_streaming_ to be false in this case. Sorry, I will fix it.

Do you think it's reasonable to let users add rate-limiting filter, otherwise, fails when the limit is reached?

Reply:
> Do you think it's reasonable to let users add rate-limiting filter, otherwise, fails when the limit is reached?

I don't think so. I think it would be more natural given the rest of the API to offer `StopAllIterationAndBuffer` and `StopAllIterationAndWatermark`, then given the return code, subsequent data can be handled in either the streaming or buffer way as the data return codes do. I don't think it would be very hard to implement. WDYT?

Reply:
I see! Is the suggestion to introduce StopAllIterationAndBuffer and StopAllIterationAndWatermark as decodeHeaders()'s return status (not as data's status)? I think it totally makes sense! Sorry for being slow!! I will change accordingly.

Reply:
Yup exactly! I think these are new headers return codes, and they guide what we do when data comes in. You aren't being slow, this is very tough stuff! :)